src/main/java/life/qbic/model/download/OutputPathFinder.java

```diff
@@ -0,0 +1,63 @@
+package life.qbic.model.download;
+
+import ch.ethz.sis.openbis.generic.dssapi.v3.dto.datasetfile.DataSetFile;
+import jline.internal.Log;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+public class OutputPathFinder {
+
+    public static Path getTopDirectory(Path path) {
+      Path currentPath = Paths.get(path.toString());
+      Path parentPath;
+      while (currentPath.getParent() != null) {
+        parentPath = currentPath.getParent();
+        currentPath = parentPath;
+      }
+      return currentPath;
+    }
+
+    public static Path determineFinalPathFromDataset(DataSetFile file, Boolean conservePaths ) {
+        Path finalPath;
+        if (conservePaths) {
+            finalPath = Paths.get(file.getPath());
+            // drop top parent directory name in the openBIS DSS (usually "/origin")
+            Path topDirectory = OutputPathFinder.getTopDirectory(finalPath);
+            finalPath = topDirectory.relativize(finalPath);
+        } else {
+            finalPath = Paths.get(file.getPath()).getFileName();
+        }
+        return finalPath;
+    }
+    private static boolean isPathValid(String possiblePath){
+        Path path = Paths.get(possiblePath);
+        return Files.isDirectory(path);
+    }
+
+    public static File determineOutputDirectoryForData(String outputPath, Path prefix, Path filePath){

```

Comment:
Here are two methods that do the same more or less.
`determineOutputDirectoryForChecksum` and `determineOutputDirectoryForData` can probably merged into one method returning a Path. Maybe you can have a look if the code calling these methods can be adapted to make this possible? It would simplify the code and would be a good refactor of the production code.